Enable data pipelining

lequytra opened the pull request:
Summary:
Note: Work in progress, publish for early feedbacks.

Currently, torchrec's built-in `TrainPipeline` API takes full control of the training loop (i.e. load next batch, forward, backward, optimizer step).

While this satisfies most common use cases, researchers who want to experiment with more advanced techniques (e.g. multi-losses, multiple batches) will have to write their own custom training loop, often with lots of duplicate codes to take advantage of comm & comp overlap speedup.

Goal of `DataPipeline` is to provide easy API to enable comm & comp overlap, but gives users back control over the training loop.

Differential Revision: D42810112

divchenko commented on torchrec/distributed/data_pipeline.py:
```diff
@@ -0,0 +1,262 @@
+    module._datapipeline._module_overlap_forward_hook()
+
+
+class DataPipeline(Iterator[In]):
```
why do we need this base class?  It's basically Iterator

lequytra replied:
Between inheriting `Generic[In]` and `Iterator[In]`, I used `Iterator[In]` for readability and enforce that `DataPipeline` stays an iterator.

divchenko commented on torchrec/distributed/data_pipeline.py:
```diff
@@ -0,0 +1,262 @@
+
+
+# pyre-ignore
+def module_overlap_forward_hook(module: torch.nn.Module, args, kwargs) -> None:
```
prefix w/ _ ?

lequytra replied:
hook is removed in updated version.

divchenko reviewed (approved):
LGTM

lequytra committed:
Enable data pipelining (#987)

Summary:
Pull Request resolved: https://github.com/pytorch/torchrec/pull/987

Note:
- Implementation in `data_pipeline.py`. Others are just refactor to share APIs between DataPipeline and TrainPipeline.

Currently, torchrec's built-in `TrainPipeline` API takes full control of the training loop (i.e. load next batch, forward, backward, optimizer step).

While this satisfies most common use cases, researchers who want to experiment with more advanced techniques (e.g. multi-losses, multiple batches) will have to write their own custom training loop, often with lots of duplicate codes to take advantage of comm & comp overlap speedup.

Goal of `DataPipeline` is to provide easy API to enable comm & comp overlap, but gives users back control over the training loop.

*Example:*
```
# vanilla training loop
for batch in data:
      # cpu model train
      optimizer.zero_grad()
      loss, pred = model_cpu(data[i])
      loss.backward()
      optimizer.step()

# old train_pipeline APIs
pipeline = TrainPipelineSparseDist(
     distributed_model, optimizer_distributed, self.device
)
for i in range(num_examples):
      pred_gpu = pipeline.progress(dataloader)

# new data_pipeline API
pipeline = SparseDistPipeline(
    dataloader_iter=dataloader, device=self.device, model=distributed_model
)
for batch in pipeline:
     optimizer_distributed.zero_grad()
     loss_pipelined, pred_pipelined = distributed_model(batch)
     loss_pipelined.backward()
      optimizer.step()
```

Reviewed By: YLGH

Differential Revision:
D42810112

Privacy Context Container: L1078771

fbshipit-source-id: b4dd8c56fc4260e8f353f8a2d96064fae17a537d
